polyhost/services/updater.py

```
@@ -207,38 +312,103 @@ def download_and_extract(tarball_url: str, tmpdir: Path,
 
-def _run_pip(args: list, label: str) -> None:
-    """Run `pip <args>` in the active interpreter; log on non-zero exit."""
+def _run_pip(args: list, label: str, line_cb=None) -> None:
+    """Run `pip <args>` in the active interpreter; log on non-zero exit.
+
+    Streams stdout+stderr line by line.  Each non-empty line is passed to
+    ``line_cb(line)`` when provided, so callers can surface it as UI feedback.
+    """
     try:
-        result = subprocess.run(
+        proc = subprocess.Popen(
```

Ruff (0.15.15)

[error] `subprocess` call: check for execution of untrusted input

(S603)